### PR TITLE
[TEST]: Regular Flows: PartialUpdate, QinQFlow, VxlanFlow specs were refactored with new approach

### PR DESCRIPTION
**Pull request by izadorozhna**
The following regular flows specs were refactored:
* PartialUpdateSpec
* QinQFlowSpec
* VxlanFlowSpec

### CODE REVIEW

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/VxlanFlowSpec.groovy [hunk 1]
```diff
@@ -380,24 +390,30 @@ class VxlanFlowSpec extends HealthCheckSpecification {
...
 
         then: "Human readable error is returned"
         def createError = thrown(HttpClientErrorException)
         createError.rawStatusCode == 400
         def createErrorDetails = createError.responseBodyAsString.to(MessageError)
         createErrorDetails.errorMessage == "Could not create flow"
-        createErrorDetails.errorDescription == getUnsupportedVxlanErrorDescription("source", switchPair.src.dpId,
-                [FlowEncapsulationType.TRANSIT_VLAN])
+        createErrorDetails.errorDescription == getUnsupportedVxlanErrorDescription(
+                "source", switchPair.src.dpId, [FlowEncapsulationType.TRANSIT_VLAN])
```
Comment (izadorozhna) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/VxlanFlowSpec.groovy:405]: @yuliiamir This version does not have the `FlowNotCreateExpectedError` comparison approach yet. Let's have a call tomorrow, I have some questions about implementation.

Comment (izadorozhna) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/VxlanFlowSpec.groovy:404]: Done

File: src-java/testing/functional-tests/src/main/groovy/org/openkilda/functionaltests/helpers/model/FlowExtended.groovy [hunk 2]
```diff
@@ -294,6 +296,11 @@ class FlowExtended {
         return new FlowExtended(flow, northbound, northboundV2, topologyDefinition, cleanupManager, database)
     }
 
+    FlowResponseV2 retrieveDetailsResponse() {
```
[the remark this answers is not in the document]
Comment (izadorozhna) [reply src-java/testing/functional-tests/src/main/groovy/org/openkilda/functionaltests/helpers/model/FlowExtended.groovy:299]: Fixed

File: src-java/testing/functional-tests/src/main/groovy/org/openkilda/functionaltests/helpers/model/FlowExtended.groovy [hunk 3]
```diff
@@ -360,11 +367,27 @@ class FlowExtended {
         return waitForBeingInState(flowState)
     }
 
+    FlowPatchV2 partialUpdateRequest() {
```
[the remark this answers is not in the document]
Comment (izadorozhna) [reply src-java/testing/functional-tests/src/main/groovy/org/openkilda/functionaltests/helpers/model/FlowExtended.groovy:370]: I've reverted the changes and now I am creating the FlowPatchV2 inside the tests and updating some fields. Please let me know if we need to apply some other approach

File: src-java/testing/functional-tests/src/main/groovy/org/openkilda/functionaltests/helpers/model/FlowExtended.groovy [hunk 4]
```diff
@@ -360,11 +367,27 @@ class FlowExtended {
...
+                destination: new FlowPatchEndpoint()
+        )
+    }
+
     FlowExtended partialUpdate(FlowPatchV2 updateRequest, FlowState flowState = FlowState.UP) {
         northboundV2.partialUpdate(flowId, updateRequest)
         return waitForBeingInState(flowState)
     }
 
+    FlowPatchDto partialUpdateRequestV1() {
```
[the remark this answers is not in the document]
Comment (izadorozhna) [reply src-java/testing/functional-tests/src/main/groovy/org/openkilda/functionaltests/helpers/model/FlowExtended.groovy:382]: +

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/PartialUpdateSpec.groovy [hunk 5]
```diff
@@ -252,30 +258,30 @@ class PartialUpdateSpec extends HealthCheckSpecification {
...
-        with(northboundV2.getFlow(flow.flowId)) {
+        with(flow.retrieveDetails()) {
             it.source.portNumber == newPortNumber
             it.source.vlanId == newVlanId
         }
 
         and: "Flow is valid and pingable"
-        northbound.validateFlow(flow.flowId).each { direction -> assert direction.asExpected }
-        with(northbound.pingFlow(flow.flowId, new PingInput())) {
+        flow.validate().each { direction -> assert direction.asExpected }
```
[the remark this answers is not in the document]
Comment (izadorozhna) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/PartialUpdateSpec.groovy:283]: Done

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/QinQFlowSpec.groovy [hunk 6]
```diff
@@ -320,80 +305,83 @@ class QinQFlowSpec extends HealthCheckSpecification {
...
         }
 
         when: "Delete the flow via APIv1"
-        northbound.deleteFlow(flow.id)
+        flow.delete()
         Wrappers.wait(WAIT_OFFSET) {
-            assert !northbound.getFlowStatus(flow.id)
-            assert northbound.getFlowHistory(flow.id).find { it.payload.last().action == DELETE_SUCCESS }
+            assert !flow.retrieveFlowStatus()
+            assert flow.retrieveFlowHistory().getEntriesByType(FlowActionType.DELETE)
```
[the remark this answers is not in the document]
Comment (izadorozhna) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/QinQFlowSpec.groovy]: Done, removed the redundant check, + found the typo that we need to use v1 API to delete the flow, but I used V2. Fixed this as well

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/PartialUpdateSpec.groovy [hunk 7]
```diff
@@ -161,24 +169,23 @@ class PartialUpdateSpec extends HealthCheckSpecification {
...
-        flowHelperV2.addFlow(flow)
-        def originalCookies = northbound.getSwitchRules(swPair.src.dpId).flowEntries.findAll { !new Cookie(it.cookie).serviceFlag }
+        def flow = flowFactory.getRandom(swPair)
+        def originalCookies = switchRulesFactory.get(swPair.src.dpId).getRules().findAll { !new Cookie(it.cookie).serviceFlag }
 
         when: "Request a flow partial update for a #data.field field"
         def newValue = data.getNewValue(flow."$data.field")
         def updateRequest = new FlowPatchV2().tap { it."$data.field" = newValue }
-        def response = flowHelperV2.partialUpdate(flow.flowId, updateRequest)
+        def response = flow.partialUpdate(updateRequest)
```
Comment (yuliiamir) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/PartialUpdateSpec.groovy:178]: flow.partialUpdate(updateRequest) includes waiting for flow to be in UP state
According to the test, we need to check response from update operation(smth like sendCreateRequest()) and then check flow details.

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/PartialUpdateSpec.groovy [hunk 8]
```diff
@@ -202,46 +209,46 @@ class PartialUpdateSpec extends HealthCheckSpecification {
...
+        def originalCookies = switchRulesFactory.get(swPair.src.dpId).getRules().findAll {
             def cookie = new Cookie(it.cookie)
             !cookie.serviceFlag || cookie.type == CookieType.MULTI_TABLE_INGRESS_RULES
         }*.cookie
 
         when: "Request a flow partial update for a 'priority' field"
         def newPriority = 777
         def updateRequest = new FlowPatchV2().tap { it.priority = newPriority }
-        def response = northboundV2.partialUpdate(flow.flowId, updateRequest)
+        def response = flow.partialUpdate(updateRequest)
```
Comment (yuliiamir) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/PartialUpdateSpec.groovy:242]: the same here about checking response from update request and then flow details

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/PartialUpdateSpec.groovy [hunk 9]
```diff
@@ -470,58 +477,48 @@ class PartialUpdateSpec extends HealthCheckSpecification {
...
         }*.cookie
 
         when: "Request a flow partial update without specifying any fields"
-        def flowBeforeUpdate = northboundV2.getFlow(flow.flowId)
-        northbound.partialUpdate(flow.flowId, new FlowPatchDto())
+        def flowBeforeUpdate = flow.retrieveDetails().convertToFlowPayload()
+        flow.partialUpdateV1(new FlowPatchDto())
 
         then: "Flow is left intact"
-        expect northboundV2.getFlow(flow.flowId), sameBeanAs(flowBeforeUpdate)
```
Comment (yuliiamir) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/PartialUpdateSpec.groovy:485]: Let's use our existing method in FlowExtended hasTheSamePropertiesAs()
 flow.partialUpdateV1(new FlowPatchDto())

then:
flow.retrieveDetails().hasTheSamePropertiesAs(flow)

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/PartialUpdateSpec.groovy [hunk 10]
```diff
@@ -470,58 +477,48 @@ class PartialUpdateSpec extends HealthCheckSpecification {
...
+        def flowBeforeUpdate = flow.retrieveDetails().convertToFlowPayload()
+        flow.partialUpdate(new FlowPatchV2())
 
         then: "Flow is left intact"
-        expect northboundV2.getFlow(flow.flowId), sameBeanAs(flowBeforeUpdate)
-                .ignoring("lastUpdated")
-                .ignoring("latencyLastModifiedTime")
-                .ignoring("forwardPathLatencyNs")
-                .ignoring("reversePathLatencyNs")
+        expect flow.retrieveDetails().convertToFlowPayload(), sameBeanAs(flowBeforeUpdate).ignoring("lastUpdated")
```
Comment (yuliiamir) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/PartialUpdateSpec.groovy:518]: the same here: hasTheSamePropertiesAs()

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/QinQFlowSpec.groovy [hunk 11]
```diff
@@ -1,78 +1,86 @@
...
 [srcVlan:#srcVlanId, srcInnerVlan:#srcInnerVlanId, dstVlan:#dstVlanId, dstInnerVlan:#dstInnerVlanId, sw:#swPair.hwSwString()]#trafficDisclaimer"() {
         when: "Create a QinQ flow"
-        def qinqFlow = flowHelperV2.randomFlow(swPair).tap {
-            source.vlanId = srcVlanId
-            source.innerVlanId = srcInnerVlanId
-            destination.vlanId = dstVlanId
-            destination.innerVlanId = dstInnerVlanId
-        }
-        def response = flowHelperV2.addFlow(qinqFlow)
+        def qinqFlow = flowFactory.getBuilder(swPair)
```
Comment (yuliiamir) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/QinQFlowSpec.groovy:58]: create() includes waiting for up state, please use sendCreateRequest()

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/QinQFlowSpec.groovy [hunk 12]
```diff
@@ -81,129 +89,115 @@ class QinQFlowSpec extends HealthCheckSpecification {
...
+        [qinqFlow, vlanFlow].each {
+            it.validateAndCollectDiscrepancies().isEmpty()
         }
 
         and: "Involved switches pass switch validation"
-        def involvedSwitchesFlow2 = pathHelper.getInvolvedSwitches(pathHelper.convert(northbound.getFlowPath(vlanFlow.id)))
-        def involvedSwitchesforBothFlows = (involvedSwitchesFlow1 + involvedSwitchesFlow2).unique { it.dpId }
-        switchHelper.synchronizeAndCollectFixedDiscrepancies(involvedSwitchesforBothFlows*.getDpId()).isEmpty()
+        def involvedSwitchesFlow2 = vlanFlow.retrieveAllEntityPaths().getInvolvedSwitches()
+        def involvedSwitchesforBothFlows = (involvedSwitchesFlow1 + involvedSwitchesFlow2).unique { it.id }
```
Comment (yuliiamir) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/QinQFlowSpec.groovy:115]: I am not sure that there is `id `param, but you can use .unique()

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/QinQFlowSpec.groovy [hunk 13]
```diff
@@ -81,129 +89,115 @@ class QinQFlowSpec extends HealthCheckSpecification {
...
-            qinqFlow.destination.vlanId = vlanFlow.destination.vlanId
-            qinqFlow.destination.innerVlanId = vlanFlow.source.vlanId
-        })
+        def updateQinqFlowEntity = qinqFlow.tap {
+            it.source.vlanId = vlanFlow.source.vlanId
+            it.source.innerVlanId = vlanFlow.destination.vlanId
+            it.destination.vlanId = vlanFlow.destination.vlanId
+            it.destination.innerVlanId = vlanFlow.source.vlanId
+        }
+        def updatedQinqFlow = qinqFlow.update(updateQinqFlowEntity)
```
Comment (yuliiamir) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/QinQFlowSpec.groovy:140]: update() includes waiting for flow to be in UP state

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/VxlanFlowSpec.groovy [hunk 14]
```diff
@@ -60,21 +65,21 @@ class VxlanFlowSpec extends HealthCheckSpecification {
...
-        flow.encapsulationType = data.encapsulationCreate
-        flowHelperV2.addFlow(flow)
+        def flow = flowFactory.getBuilder(swPair)
+                .withEncapsulationType(data.encapsulationCreate)
+                .build().create()
 
         then: "Flow is created with the #encapsulationCreate.toString() encapsulation type"
-        def flowInfo = northboundV2.getFlow(flow.flowId)
-        flowInfo.encapsulationType == data.encapsulationCreate.toString().toLowerCase()
+        def flowInfo = flow.retrieveDetails()
```
Comment (yuliiamir) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/VxlanFlowSpec.groovy:73]: can omit this call as we are waiting for flow to be in UP state and it is the call flow.retrieveDetails()

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/VxlanFlowSpec.groovy [hunk 15]
```diff
@@ -540,24 +552,9 @@ class VxlanFlowSpec extends HealthCheckSpecification {
 
         where:
         encapsulationCreate                | encapsulationUpdate
-        FlowEncapsulationType.TRANSIT_VLAN | VXLAN
-        VXLAN | FlowEncapsulationType.TRANSIT_VLAN
-
-    }
+        FlowEncapsulationType.TRANSIT_VLAN | FlowEncapsulationType.VXLAN
+        FlowEncapsulationType.VXLAN | FlowEncapsulationType.TRANSIT_VLAN
```
Comment (yuliiamir) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/VxlanFlowSpec.groovy:556]: minor: reformat code

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/PartialUpdateSpec.groovy (hunk 7 above)
Comment (izadorozhna) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/PartialUpdateSpec.groovy:178]: Done

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/PartialUpdateSpec.groovy (hunk 8 above)
Comment (izadorozhna) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/PartialUpdateSpec.groovy:242]: Done

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/PartialUpdateSpec.groovy (hunk 9 above)
Comment (izadorozhna) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/PartialUpdateSpec.groovy:485]: Done

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/PartialUpdateSpec.groovy (hunk 10 above)
Comment (izadorozhna) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/PartialUpdateSpec.groovy:518]: Done

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/QinQFlowSpec.groovy (hunk 11 above)
Comment (izadorozhna) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/QinQFlowSpec.groovy:58]: Done

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/QinQFlowSpec.groovy (hunk 12 above)
Comment (izadorozhna) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/QinQFlowSpec.groovy:115]: Done. Yes, there was `id` field there, however .unique() is simplier

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/QinQFlowSpec.groovy (hunk 13 above)
Comment (izadorozhna) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/QinQFlowSpec.groovy:140]: Done, fixed to sendUpdateRequest()

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/VxlanFlowSpec.groovy (hunk 14 above)
Comment (izadorozhna) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/VxlanFlowSpec.groovy:73]: Done, removed

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/VxlanFlowSpec.groovy (hunk 15 above)
Comment (izadorozhna) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/VxlanFlowSpec.groovy:556]: Done

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/QinQFlowSpec.groovy [hunk 16]
```diff
@@ -1,209 +1,206 @@
...
-        with(response) {
+        with(qinqFlow) {
             it.source.vlanId == srcVlanId
             it.source.innerVlanId == srcInnerVlanId
             it.destination.vlanId == dstVlanId
             it.destination.innerVlanId == dstInnerVlanId
         }
 
         and: "Flow is really created with requested vlanIds"
-        with(northbound.getFlow(qinqFlow.flowId)) {
```
[the remark this answers is not in the document]
Comment (izadorozhna) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/QinQFlowSpec.groovy:66]: Done

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/PartialUpdateSpec.groovy [hunk 17]
```diff
@@ -1,80 +1,89 @@
...
         when: "Request a flow partial update for a #data.field field"
         def updateRequest = new FlowPatchV2().tap { it."$data.field" = data.newValue }
-        def response = northboundV2.partialUpdate(flow.flowId, updateRequest)
+        def response = flow.sendPartialUpdateRequest(updateRequest)
 
         then: "Update response reflects the changes"
         response."$data.field" == data.newValue
 
         and: "Changes actually took place"
         Wrappers.wait(WAIT_OFFSET) {
```
[the remark this answers is not in the document]
Comment (izadorozhna) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/PartialUpdateSpec.groovy:77]: Done

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/QinQFlowSpec.groovy [hunk 18]
```diff
@@ -320,80 +310,85 @@ class QinQFlowSpec extends HealthCheckSpecification {
...
 
         then: "Partial update response reflects the changes"
         response.destination.vlanId == newDstVlanId
         response.destination.innerVlanId == newDstInnerVlanId
 
         and: "Flow is really updated with requested innerVlanId/vlanId"
-        with(northbound.getFlow(flow.id)) {
-            it.destination.vlanId == newDstVlanId
-            it.destination.innerVlanId == newDstInnerVlanId
+        Wrappers.wait(FLOW_CRUD_TIMEOUT) {
```
[the remark this answers is not in the document]
Comment (izadorozhna) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/QinQFlowSpec.groovy:362]: +

**Review by yuliiamir — APPROVED** [review]

**Comment by pablomuri** [comment]
Hi, could you please fix the conflict?

**Pull request merged**